client/components/localized-moment/context.js

```diff
@@ -19,13 +19,13 @@ const debug = debugFactory( 'calypso:localized-moment' );
 const MomentContext = React.createContext( moment );
 
 class MomentProvider extends React.Component {
-	state = { moment, momentLocale: 'en' };
+	state = { moment, momentLocale: moment.locale() };
 
-	async checkAndLoad( previousLocale ) {
+	async checkAndLoad() {
 		const { currentLocale } = this.props;
 
 		// has the requested locale changed?
-		if ( currentLocale === previousLocale ) {
+		if ( currentLocale === this.state.momentLocale ) {
```

Comment:
There was a reason why the comparison was done against `prevProps.currentLocale` rather than `this.state.momentLocale`.

Both values can be out of sync while the locale data are loaded from network. Sequence of events:
1. Rerender provider with new `currentLocale="de"` (previous value was "cs").
2. Request for the `moment-locale-de` chunk is issued.
3. Consumers stay rendered with `momentLocale="cs"`, no update yet.
4. Request for `moment-locale-de` chunk finishes, `state.momentLocale` gets updated from `cs` to `de`, consumers rerendered.

After the patch, there is a race condition:
1. Provider rendered with `currentLocale="de"`
2. Provider rerendered with `currentLocale="cs"`
3. Request for `moment-locale-cs` is issued
4. While request is still in flight, rerenderer with `currentLocale="de"` (back to original value)
5. Because new prop is equal to current state (`de`), no network request is issued
6. Request for `moment-locale-de` finishes, sets locale to `cs`. Not wanted!

Reply:
I can try to make this component robust against race conditions, but that will take significantly more work than what the previous implementation was doing. As far as I can tell, it was already vulnerable to requests arriving out of order, for example.

Reply:
The previous implementation was vulnerable, too, that's true 🙂 But I don't think it should be very hard to make it robust. The main principles to implement are:
- there is "desired locale" (from prop) and "actual locale" (in state and in `moment.locale`)
- these can be out of sync while the network request is in flight: the desired locale doesn't become the actual one immediately, but only after a while
- when network response arrives, we should check that the response is still wanted. The current implementation doesn't do that.

It should be a one-line fix? Checking the current value if `this.props.currentValue` after the `await` finishes?

By the way, why did you need to change the `checkAndLoad` method at all? If the state defaulted to `moment.locale()` instead of `en`, and the `checkAndLoad` check in didMount was called with `this.state.momentLocale`, wouldn't that be a sufficient fix to avoid the rerender? Use cases to test:
- initial mount with `en` does nothing, dates formatted in `en`.
- initial mount with `cs` issues locale load request, dates formatted in `en` for a second, then reformatted to `cs`
- rerender from `cs` to `de` loads the `de` locale and rerenders.
- mounting a new provider immediately renders its consumers with the current locale, no requests issued

> Modules are spec'ed to be deduplicated

That's right, no extra request will be issued. I forgot it's a webpack chunk. :+1:

Thanks for finding and fixing the rerender issue anyway! The edge cases are very subtle and not very urgent to fix.

Reply:
> By the way, why did you need to change the `checkAndLoad` method at all? If the state defaulted to `moment.locale()` instead of `en`, and the `checkAndLoad` check in didMount was called with `this.state.momentLocale`, wouldn't that be a sufficient fix to avoid the rerender?

It was mostly to avoid having the default state duplicated, but the approach was somewhat over-the-top. I ended up rewriting some complicated logic just to avoid that. Duly reverted 🙂 

> It should be a one-line fix? Checking the current value if this.props.currentValue after the await finishes?

I suspect that doesn't cover everything (e.g. out-of-order messages with multiple changes between locales), but it does make it much less likely to happen. I'll add a comment so that it's clear what's going on.

Thanks for the thorough review as usual, @jsnajdr! No edge case gets past you! 😄

Filed #31691 with the fixes.